php/class-asset-manager-preload.php

```diff
@@ -0,0 +1,208 @@
+<?php
+/**
+ * Class file for Asset_Manager_Preload
+ *
+ * @package AssetManager
+ */
+
+/**
+ * Asset_Manager_Preload class.
+ */
+class Asset_Manager_Preload extends Asset_Manager {
+
+	/**
+	 * Holds references to the singleton instances.
+	 *
+	 * @var array
+	 */
+	private static $instance;
+
+	/**
+	 * Types of files that can be preloaded; corresponds to allowed `as` attribute values.
+	 *
+	 * @var array
+	 */
+	public $preload_as = [
+		'audio',
+		'document',
+		'embed',
+		'fetch',
+		'font',
+		'image',
+		'object',
+		'script',
+		'style',
+		'track',
+		'worker',
+		'video',
+	];
+
+	/**
+	 * Asset type this class is responsible for loading and managing.
+	 *
+	 * @var string
+	 */
+	public $asset_type = 'preload';
+
+
+	/**
+	 * Methods by which an asset can be loaded into the DOM.
+	 *
+	 * @var array
+	 */
+	public $load_methods = [ 'preload' ];
+
+	/**
+	 * Map of asset 'as' and 'type` attributes based on file extension, used to
+	 * patch in attributes for commonly-preloaded assets.
+	 * https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
+	 *
+	 * @var array
+	 */
+	public $asset_types = [
+		'css'   => [
+			'as'        => 'style',
+			'mime_type' => 'text/css',
+		],
+		'js'    => [
+			'as'        => 'script',
+			'mime_type' => 'text/javascript',
+		],
+		'woff2' => [
+			'as'        => 'font',
+			'mime_type' => 'font/woff2',
+		],
+	];
+
+	/**
+	 * Constructor.
+	 */
+	private function __construct() {
+		// Don't do anything, needs to be initialized via instance() method.
+	}
+
+	/**
+	 * Get an instance of the class.
+	 *
+	 * @return class
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new static();
+			self::$instance->add_hooks();
+			self::$instance->set_defaults();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Print a single asset
+	 *
+	 * @param array $asset Asset to insert into DOM.
+	 *
+	 * @return void
+	 */
+	public function print_asset( $asset ) {
+		$classes      = $this->default_classes;
+		$classes[]    = $asset['handle'];
+		$print_string = '';
+
+		if ( empty( $asset['as'] ) || ! in_array( $asset['as'], $this->preload_as, true ) ) {
+			// We weren't able to patch in the 'as' attribute in `post_validate_asset`.
+			$this->generate_asset_error( 'invalid_preload_attribute', $asset, [ 'attribute' => 'as' ] );
```

Comment:
How about adding this error code to the function generate_asset_error so that a specific error is displayed?

Reply:
Hmm... I thought I did. Thanks! I'll double-check to make sure.